### PR TITLE
Feature 29 extensible discovery

### PR DESCRIPTION
**Pull request by bechte**
Resolved and fixed all issues from #29 and #39.

Possible improvements that we can work on after this merge:
- [ ] Define a concept of a "path" along the unique id
- [ ] Use "path" concept to
  - [ ] remove tight coupling between resolvers
  - [ ] split `ClassResolver` into two classes: `TopLevelClassResolver` & `NestedClassResolver`
  - [ ] extract `NestedClassResolver` & `NestedClassTestDescriptor` as an extension
- [ ] Clarify whether or not to keep static inner classes on the top level
- [ ] Rework the `TestPlanSpecification`: Extensibility vs. Data Structure

I would like to handle those improvements if necessary in separate tickets working on the master branch. This will allow us to work in parallel on the topics.

### CODE REVIEW

File: junit-commons/src/main/java/org/junit/gen5/commons/util/ObjectUtils.java [hunk 1]
```diff
@@ -29,4 +31,8 @@ public static String nullSafeToString(Class<?>... classes) {
 		return stream(classes).map(Class::getName).collect(joining(", "));
 	}
 
+	public static void verifyNonNull(Object obj, String message) {
+		if (obj == null)
+			throw new IllegalArgumentException(message);
+	}
```
Comment (marcphilipp) [inline junit-commons/src/main/java/org/junit/gen5/commons/util/ObjectUtils.java]: This is exactly the same as `Preconditions.notNull()`. Please remove this method again and use the existing one instead. 😉

File: junit-tests/src/test/java/org/junit/gen5/engine/DummyTestEngineDescriptor.java [hunk 2]
```diff
@@ -10,20 +10,21 @@
 
 package org.junit.gen5.engine;
 
-class DummyTestDescriptor extends AbstractTestDescriptor implements Leaf<DummyEngineExecutionContext> {
-
-	public static final String ENGINE_ID = "dummy";
-
+class DummyTestEngineDescriptor extends AbstractTestDescriptor implements Leaf<DummyEngineExecutionContext> {
```
Comment (marcphilipp) [inline junit-tests/src/test/java/org/junit/gen5/engine/DummyTestEngineDescriptor.java]: Why did you rename this class? I think the new name is confusing because it's not actually an `EngineDescriptor`.

Comment (bechte) [inline junit-tests/src/test/java/org/junit/gen5/engine/DummyTestEngineDescriptor.java]: Based on the unique id of this dummy, it represents a test engine node. That is why I renamed it. I agree the new name does not reflect it either. `implements Leaf<bla>` was added after I changed the name, so the intent of this dummy might have changed. I can rename it back to what it was. Still I think we should talk about a naming convention for test doubles.

File: junit-commons/src/main/java/org/junit/gen5/commons/util/ObjectUtils.java (hunk 1 above)
Comment (bechte) [inline junit-commons/src/main/java/org/junit/gen5/commons/util/ObjectUtils.java]: True :) - it was late. Thanks.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java [hunk 3]
```diff
@@ -57,6 +54,17 @@
...
 	@Override
+	public String getUniqueId() {
+		// @formatter:off
+		return String.format("%s:%s#%s(%s)",
+				getTestEngine().getId(),
+				getTestClass().getName(),
+				getTestMethod().getName(),
+				ObjectUtils.nullSafeToString(getTestMethod().getParameterTypes()));
+		// @formatter:on
+	}
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java]: Is this correct for methods in nested non-static classes?

Here and in the other `getUniqueId()` methods: Have you considered implementing this like the following (pseudo-code)?
```java
@Override
public String getUniqueId() {
	// @formatter:off
	return String.format("%s#%s(%s)",
			getParent().map(TestDescriptor::getUniqueId).orElse("")
			getTestMethod().getName(),
			ObjectUtils.nullSafeToString(getTestMethod().getParameterTypes()));
	// @formatter:on
}
```

File: junit-tests/src/test/java/org/junit/gen5/engine/DummyTestEngineDescriptor.java (hunk 2 above)
Comment (marcphilipp) [inline junit-tests/src/test/java/org/junit/gen5/engine/DummyTestEngineDescriptor.java]: Well, at the moment it is used for the engine node and its children. The unique IDs are build like this: `ENGINE_ID + ":someString"`. I agree that this could be cleaned up, but would prefer the original class name.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java (hunk 3 above)
Comment (bechte) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/descriptor/MethodTestDescriptor.java]: Yep, it is correct (we have tests for this). Anyways, I am with you, we should investigate some more time on this. I would like to have a separate ticket for it, as it is not as easy as it looks in the first place. If you like, we could pair on this topic and have a look at your "path" idea.

I have considered to create the unique id using the parent, but in my first version this did not work out, as I moved methods from the specification on the top level. After we decided to create the whole tree for every node, it is now possible to use the parent for id generation. As I said, I would like to consider this topic in an own issue, as I think it needs some more time to find out all dependencies.

File: junit5-engine/src/main/java/org/junit/gen5/engine/junit5/resolver/TestResolver.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+package org.junit.gen5.engine.junit5.resolver;
+
+import org.junit.gen5.engine.TestDescriptor;
+import org.junit.gen5.engine.TestEngine;
+import org.junit.gen5.engine.TestPlanSpecification;
+
+/**
+ * A {@link TestResolver} is responsible for resolving different kind of test representatives. Each resolver can
+ * contribute to the list of children of the given parent. It may only resolve children, that accomplish the
+ * {@link TestPlanSpecification}. The children are returned as list.
```
Comment (marcphilipp) [inline junit5-engine/src/main/java/org/junit/gen5/engine/junit5/resolver/TestResolver.java]: I believe the last sentence is outdated.

**Comment by bechte** [comment]
This feature will be discussed internally and rebuilt onto master. Therefore, I am closing this pull request.
